Handle blocks out of topological order

cpacia opened the pull request:
This commit refactors the block scanning algorithm to handle checking for relevant transactions when the block is not in topological order (for example, when CTOR is active).

We do this by keeping a map of all of the inputs in a block. Whenever a tx is processed, we check its 
hash against the input map, if there is a match we process that previous transaction again and likewise recursively process it's dependencies.

This is very complex to reason about and I'm not sure if I did it 100% correct. Needs solid review.

closes #6

cpacia committed:
Handle blocks out of topological order

This commit refactors the block scanning algorithm to handle checking for
relevant transactions when the block is not in topological order (for example,
when CTOR is active).

We do this by keeping a map of all of the inputs in a block. Whenever a tx
is processed, we check its hash against the input map, if there is a match
we process that previous transaction again and likewise recursively process
its dependencies.

cpacia commented on chain/block_filterer.go:
```diff
@@ -100,21 +101,32 @@ func NewBlockFilterer(params *chaincfg.Params,
-			bf.RelevantTxns = append(bf.RelevantTxns, tx)
-			hasRelevantTxns = true
+		for _, in := range tx.TxIn {
+			inputs[in.PreviousOutPoint.Hash] = tx
```
what if two txs spend from the same input tx?

zquestz replied:
Yeah I am not sure that will work, we should definitely write a test for this case.

cpacia replied:
Maybe it should map to a list of txs?

zquestz replied:
Yeah I think a slice of tx's would be needed. However, we will know if we add the required test for this case. I have a feeling it will be useful to keep around if we ever refactor this code. =)

zquestz commented on chain/block_filterer_test.go:
```diff
@@ -300,6 +461,45 @@ func TestBlockFiltererOneInOneOut(t *testing.T) {
+	}
+	blockFilterer := chain.NewBlockFilterer(&chaincfg.SimNetParams, req)
+
+	// Filter block 100000, which should find matches for the watched
```
NIT: isn't this block 200000?

cpacia committed:
Update input map to use list of transactions

cpacia commented:
Updated the PR to map the input hash to a slice of transactions. Also added another transaction to the test case which spends from the second output of the receiving transaction.

zquestz reviewed (approved):
Going to assume you ran the test on the old code and that it failed. This change looks good to me. =)
